File: day1/cplusplus/main.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>
#include <stdexcept>

using namespace std;

pair<vector<int>, vector<int>> read_input(const string &filename = "../../input.txt")
{
    ifstream file(filename);
    if (!file)
    {
        throw runtime_error("Cannot open input file: " + filename);
    }

    vector<int> col1, col2;
    int a, b;
    while (file >> a >> b)
    {
        col1.push_back(a);
        col2.push_back(b);
    }
    return {col1, col2};
}

void part1()
{

    // WHY AUTO?
    // tells compiler to automatically deduce the type
    // but this removes type safety...
    // so better not use
    pair<vector<int>, vector<int>> columns = read_input();
    vector<int> sorted1 = columns.first;
    vector<int> sorted2 = columns.second;

    sort(sorted1.begin(), sorted1.end());
    sort(sorted2.begin(), sorted2.end());

    int sum = 0;
    for (size_t i = 0; i < sorted1.size(); ++i)
    {
        sum += abs(sorted1[i] - sorted2[i]);
    }

    cout << "PART 1\nSOLUTION: " << sum << "\n\n";
}

void part2()
{
    pair<vector<int>, vector<int>> columns = read_input();
    vector<int> c1 = columns.first;
    vector<int> c2 = columns.second;

    int score = 0;
    for (const int x : c1)
    {
        score += x * count(c2.cbegin(), c2.cend(), x);
    }

    cout << "PART 2\nSOLUTION: " << score << "\n\n";
}

int main()
{
    try
    {
        cout << "Advent of Code 2024 - Day 1\n\n";
        part1();
        part2();
        return 0;
    }
    catch (const exception &e)
    {
        cerr << "Error: " << e.what() << '\n';
        return 1;
    }
}

File: day1/python/main.py
def read_input(filename="../input.txt"):
    # Step 1 - read the file and split into columns
    with open(filename, 'r') as file:
        lines = file.readlines()
        column1 = []
        column2 = []
        for line in lines:
            col1, col2 = line.split()
            column1.append(int(col1))
            column2.append(int(col2))

    return column1, column2


def part1():

    column1, column2 = read_input()
    
    # Step 2 - sort the columns
    column1.sort()
    column2.sort()

    # Step 3 - Calculate the sum of the differences
    sum = 0
    for i in range(len(column1)):
        sum += abs(column1[i] - column2[i])

    print("PART 1")
    print("SOLUTION: ", sum)
    print()

def part2(): 

    column1, column2 = read_input()

    # Step 1 - calculate similarity score
    similarity_score = 0
    for element in column1:
        similarity_score += element*column2.count(element)


    print("PART 2")
    print("SOLUTION: ", similarity_score)
    print()

    # Calculate similarity 
if __name__ == '__main__':

    print("Advent of Code 2024 - Day 1\n")

    ##
    part1()
    ## 
    part2()